### PR TITLE
Week5 tdd

### PR DESCRIPTION
**Pull request by ERetana2**

### CODE REVIEW

File: portfolio/src/main/webapp/script.js [hunk 1]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//Set of Colors for map
+// Set of Colors for map
```
Comment (qihuang2) [inline portfolio/src/main/webapp/script.js]: these don't seem related to the TDD stuff. we should create a separate PR

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java [hunk 2]
```diff
@@ -16,6 +16,7 @@
 
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Comparator;
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java]: this change doesn't seem to be needed since you're not making any other changes to the file

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,48 @@
...
 
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:41]: add javadoc for what this method does: https://google.github.io/styleguide/javaguide.html#s7-javadoc

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,48 @@
...
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    ArrayList<TimeRange> unavailableTimes = new ArrayList<>();
+    int meetingStartTime = TimeRange.START_OF_DAY;
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: move the declaration down to where it's being used. it makes code more readable when the variables you use are declared where you use them

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,48 @@
...
+    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+
+    for(TimeRange time: unavailableTimes){
+      if(time.contains(meetingStartTime)){
+        meetingStartTime = time.end();
+      }else{
+        if((time.start() - meetingStartTime) >= duration){
+          availableTimes.add(TimeRange.fromStartEnd(meetingStartTime, time.start(), false));
+        }
+        if(meetingStartTime > time.end()){break;}
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: format to 
if(meetingStartTime > time.end()){
  break;
}

https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

i thinking running the formatter will fix all these

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,48 @@
...
+      if(!Collections.disjoint(event.getAttendees(), request.getAttendees())){
+        unavailableTimes.add(event.getWhen());
+      }
+    }
+    //Sort the list of unavailable times and check for valid available time
+    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+
+    for(TimeRange time: unavailableTimes){
+      if(time.contains(meetingStartTime)){
+        meetingStartTime = time.end();
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: wouldn't you want the greater value? meetingStartTime = Math.max(time.end(), meetingStartTime)

**Review by qihuang2 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 1 above)
Comment (qihuang2) [reply portfolio/src/main/webapp/script.js]: if you created a separate PR for these changes, you should do it for this PR

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,87 @@
...
+    ArrayList<TimeRange> empty = new ArrayList<>();
+    ArrayList<String> attendees = new ArrayList<String>(request.getAttendees()); // list for mandatory attendees
+    ArrayList<String> allAttendees = new ArrayList<String>(attendees);
+    allAttendees.addAll(request.getOptionalAttendees());
+    // return whole day if there are no attendees
+    if (allAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } 
+    //return empty If there are no events during day(gaps), but there are optional attendees
+    if(events.isEmpty() && !request.getOptionalAttendees().isEmpty()){
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: if events are empty, wouldn't the whole day be free then?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,87 @@
...
 public final class FindMeetingQuery {
+  /**
+   * @param events a collection of events going on during the day
+   * @param request user created request for an event that includes Time and attendees
+   * @return a list of available times to schedule a meeting
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> empty = new ArrayList<>();
+    ArrayList<String> attendees = new ArrayList<String>(request.getAttendees()); // list for mandatory attendees
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: rename attendees to mandatoryAttendees. then there is no need for "// list for mandatory attendees" comment since variable name is self explanatory

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,87 @@
...
+    if(events.isEmpty() && !request.getOptionalAttendees().isEmpty()){
+      return empty;
+    }
+    // When there are no available times with optional attendees, set available times for mandatory employees only
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>(
+        availableTimesForAttendees(events, allAttendees, request.getDuration()));
+    if(!availableTimes.isEmpty()){
+      return availableTimes;
+    }
+    else{
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: else should be on same line as closing brace for if:
"} else {"

https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

fix here and rest of file

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,87 @@
...
+      availableTimes = new ArrayList<TimeRange>(availableTimesForAttendees(events, attendees, request.getDuration()));
+    }
+    // If the request has no attendees after optional attendees are excluded, return an empty collection.
+    if(!attendees.isEmpty()){
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: can't we check for attendees empty before running line "availableTimes = new ArrayList<TimeRange>(availableTimesForAttendees(events, attendees, request.getDuration()));"

that will prevents unnecessary calls to availableTimesForAttendees

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,87 @@
...
+    int meetingStartTime = TimeRange.START_OF_DAY;
+    for(TimeRange time: unavailableTimes){
+      if(time.contains(meetingStartTime)){
+        meetingStartTime = Math.max(time.end(),meetingStartTime);
+      }else{
+        if((time.start() - meetingStartTime) >= duration){
+          availableTimes.add(TimeRange.fromStartEnd(meetingStartTime, time.start(), false));
+        }
+        if(meetingStartTime > time.end()){
+          break;
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: did you want a break here? break will stop the for loop. did you want continue?

**Review by qihuang2 — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (ERetana2) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: according to the tutorial, no if there are optional attendees included to the meeting request

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -204,4 +204,4 @@ function initMap() {
   marker.addListener('click', function() {
     infowindow.open(map, marker);
   });
-}
+}
```
Comment (qihuang2) [inline portfolio/src/main/webapp/script.js]: nit: add new line at end of file

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -14,10 +14,85 @@
...
 
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  /**
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:25]: can you add some more comments about the edge cases? like if there are no mandatory attendees but are optional attendees, what happens? should also mention that optional attendees are first considered when finding available times, but they are removed from consideration if no available times are found with them considered

javadocs are useful for people that might come along later, and have no idea what this method is supposed to do. they can read the javadoc and know what the method promises to do without having to read the source code. stating what happens during edge cases is extremely helpful

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -14,10 +14,85 @@
...
+    //return empty If there are no events during day(gaps), but there are optional attendees
+    if(events.isEmpty() && !request.getOptionalAttendees().isEmpty()){
+      return empty;
+    }
+    // When there are no available times with optional attendees, set available times for mandatory employees only
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>(
+        availableTimesForAttendees(events, allAttendees, request.getDuration()));
+    if(!availableTimes.isEmpty()){
+      return availableTimes;
+    }else{
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:58]: nit: you can get rid of this else since you're returning early in if(!availableTimes.isEmpty()){
      return availableTimes;
    }

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -14,10 +14,85 @@
...
+      // If the request has no attendees after optional attendees are excluded, return an empty collection.
+      if(!mandatoryAttendees.isEmpty()){
+        availableTimes = new ArrayList<TimeRange>(availableTimesForAttendees(events, mandatoryAttendees, request.getDuration()));
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: you can do 
return new ArrayList<TimeRange>(availableTimesForAttendees(events, mandatoryAttendees, request.getDuration()));

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -14,10 +14,85 @@
...
 
 public final class FindMeetingQuery {
+  /**
+   * @param events a collection of events going on during the day
+   * @param request user created request for an event that includes Time and attendees
+   * @return a list of available times to schedule a meeting
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> empty = new ArrayList<>();
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: it's better not to save this as a variable. just do return new ArrayList<>() where it's needed. Instantiating objects aren't free, and there's no need to instantiate this unless it's actually going to be used

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -14,10 +14,85 @@
...
+      if(time.contains(meetingStartTime)){
+        meetingStartTime = Math.max(time.end(),meetingStartTime);
+      }else{
+        if((time.start() - meetingStartTime) >= duration){
+          availableTimes.add(TimeRange.fromStartEnd(meetingStartTime, time.start(), false));
+        }
+        if(meetingStartTime > time.end()){
+          continue;
+        }
+        meetingStartTime = time.end();
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: do 
if(meetingStartTime < time.end()) {
   meetingStartTime = time.end();
}
here instead. it's a little more readable

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 18]
```diff
@@ -121,6 +123,66 @@ public void everyAttendeeIsConsidered() {
...
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+    System.out.println();
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: get rid of this println

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 19]
```diff
@@ -121,6 +123,66 @@ public void everyAttendeeIsConsidered() {
...
+  }
+
+  @Test
+  public void everyAttendeeIsConsideredV3() {
+    // Have each person have different events. We should see two options because each person has
+    // split the restricted times.
+    //
+    // Events  :       |--A--|     |--B--|
+    // Optional: |--------------C--------------|
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: this graphic doesn't match the code

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 20]
```diff
@@ -270,5 +360,49 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
+
+  @Test
+  public void noneMandatory(){
+    //Have no mandatory attendees, Just 2 optional
+    //
+    // Events  :
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:368]: add optional events

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 21]
```diff
@@ -270,5 +360,49 @@ public void notEnoughRoom() {
...
+  public void noneMandatoryNoGaps(){
+    //Have no mandatory attendees, Just 2 optional
+    //
+    // Events  :
+    // Day     : |--------------------|
+
+    Collection<Event> events = Collections.emptySet();
+
+    MeetingRequest request = new MeetingRequest(new ArrayList<>(), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:407]: no gaps, so would the optionals have events for the whole day?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (qihuang2) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: add this edge case to javadoc

**Review by qihuang2 — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 21 above)
Comment (ERetana2) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:407]: added to javadoc for clarification, no gaps refers to no events for attendees

Comment (qihuang2) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:407]: I'm not sure if that makes sense then. If there are optional attendees, but the attendees are free all day, wouldn't we get the WHOLE_DAY as free?

Scenario 2: When there is no mandatory attendees and there is optional attendees with gaps
   * return available times around gaps for optional attendees

The gaps for optional attendees will be the whole day

**Review by qihuang2 — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 21 above)
Comment (qihuang2) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:407]: i think "gap" is referring to when they are free

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -14,10 +14,98 @@
...
+   * return available times for mandatory attendees
+   * Scenario 5: When there is mandatory attendees and there is optional attendees with gaps
+   * return times available for both -> if there are no times available for both return
+   * available times for mandatory attendees
+   * Scenario 6: When there is mandatory attendees with no events and no optional attendees
+   * return the whole day
+   * Scenario 7: When there is mandatory attendees with no events and optional attendees with events
+   * return gaps in optional attendees
+   * Scenario 8: When there is mandatory attendees with no events and optional attendees with events
+   * return the whole day
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: move this to the start of javadoc (before @param)

**Review by qihuang2 — CHANGES_REQUESTED** [review]

**Comment by ERetana2** [comment]
fixed misinterpreted scenario! thanks !

**Review by qihuang2 — APPROVED** [review]
Looks great!
